### DISCUSSION TITLE
Request-scoped state for interceptor on Kafka message processor

### DISCUSSION THREAD

**Discussion by knutwannheden** [quarkusio/quarkus#28124] [Q&A] [answered]
I have an application-scoped Kafka message processor (with an `@Incoming` annotated method) which ends up calling a custom CDI interceptor bound to the `@Incoming` annotation. Inside this interceptor I inject an `@RequestScoped` bean, because I use it to store some request-specific data (a request here would correspond to a Kafka message).

This used to work well in Quarkus 2.12, but in Quarkus 2.13 I now get an error logged advising me to add an `@ActivateRequestContext` annotation. Doing that on the Kafka message processor doesn't appear to work, however. I haven't had time to analyze where things go wrong.

Before reporting an issue I would like to check if this setup even makes sense. Is there some better way to store request-scoped data other than using a stateful request-scoped bean?

Background: I want to transparently extract some headers from the Kafka message and have the values available to the processing code for the scope of the Kafka message processing. It would also need to work with context propagation.

/cc @ozangunalp @cescoffier

[a comment by an automated account is not shown]

**Comment by cescoffier** [comment, marked as answer]
So, first, it was not working before, as all the messages had the same request scope (attached to the whole stream and not to messages). To avoid confusion, we fixed that, and now it must be explicitly activated using `@ActivateRequestContext`. However, the request scope will be terminated when the method returns. 

To do the extraction and store the data, you will need to use the duplicated context associated with the message or the message metadata. Your interceptor is called on a duplicated context that is associated with the message. So, the next processing stage of that message will be executed in the same context. You can use `ContextLocals` in your interceptor code to store and restore the data.

**Reply by gsmet** [reply]
I think we need a significant migration guide entry for all these changes. As they are far from obvious.

**Reply by cescoffier** [reply]
It was totally broken before, so, it's like a migration from something broken to something that works.

**Reply by knutwannheden** [reply]
Also, it would be very helpful with some practical advice on how to implement a state that can be used for both blocking and non-blocking code. I.e. how at the beginning and at the end of the call / processing this state can be set and then again cleared. It would also have to support context propagation in both cases.

**Comment by ozangunalp** [comment]
I'd like to add, a request doesn't correspond to a Kafka message, as a request's scope is terminated by the response, whereas it is difficult to demarcate the scope of processing of a message. If your processing is limited to the `@Incoming` method, you may use `@ActivateRequestContext`, as Clement said the scope will be terminated when the method (or the returned Uni/CompletionStage) completes. The `ActivateRequestContext` is already an interceptor, If you are using an interceptor make sure it is called after, `ActivateRequestContextInterceptor` priority is `100`.

If you use the duplicated context to store data, and here is a usage of `ContextLocals` : https://github.com/quarkusio/quarkus/blob/main/integration-tests/reactive-messaging-hibernate-reactive/src/main/java/io/quarkus/it/kafka/KafkaReceivers.java#L33

**Reply by knutwannheden** [reply]
I was also going to suggest that a user guide would be really helpful, ad I haven't come across any documentation yet.

**Reply by ozangunalp** [reply]
There is one here : https://github.com/quarkusio/quarkus/wiki/Migration-Guide-2.13#smallrye-reactive-messaging but maybe we need to elaborate more.

**Reply by knutwannheden** [reply]
I had missed that, thanks. But what I meant was that I think it would be nice with some documentation regarding when and how to use duplicated context and `ContextLocals`.
